test/benchmark/iam/LanguageFeatureTest.java

```diff
@@ -0,0 +1,179 @@
+                "$p (subject: $s, access: $ac) isa permission;\n";
+        Benchmark benchmark = new Benchmark("relation-with-role-players", query, 10);
+        benchmarker.runBenchmark(benchmark);
+
+        benchmark.assertAnswerCountCorrect();
+        benchmark.assertRunningTime(1000);
+        benchmark.assertCounters(500, 22, 30, 72);
+    }
+
+    @Test
+    public void testRelationWithoutRolePlayer() {
+        String query = "match\n" +
+                "$p isa permission;\n";
+        Benchmark benchmark = new Benchmark("relation-without-role-players", query, 2783);
+        benchmarker.runBenchmark(benchmark);
+
+        benchmark.assertAnswerCountCorrect();
+        benchmark.assertRunningTime(5000);
+        benchmark.assertCounters(100, 2894, 800, 2500);
+    }
+
+    @Test
+    public void testInferredRelationAndOwnership() {
+        String query = "match\n" +
+                "$p (subject: $s, access: $ac) isa permission, has validity $v;\n";
+        Benchmark benchmark = new Benchmark("inferred-relation-and-ownership", query, 2783);
+        benchmarker.runBenchmark(benchmark);
+
+        benchmark.assertAnswerCountCorrect();
+        benchmark.assertRunningTime(10_000);
+        benchmark.assertCounters(500, 5677, 10000, 15000);
+    }
+    // Less common features
+
+    @Test
+    public void testBoundRelation() {
+        String query = "match\n" +
+                "$p isa person, has email \"douglas.schmidt@vaticle.com\";\n" +
+                "$f isa file, has path \"root/engineering/typedb-studio/src/README.md\";\n" +
+                "$o isa operation, has name \"edit file\";\n" +
+                "$a (object: $f, action: $o) isa access;\n" +
+                "$pe (subject: $p, access: $a) isa permission;\n" +
+                "$pe is $pe-same;\n" +
+                "$pe-same (subject: $other-p, access: $other-a) isa permission;\n";
+        Benchmark benchmark = new Benchmark("bound-relation", query, 1);
+        benchmarker.runBenchmark(benchmark);
+
+        benchmark.assertAnswerCountCorrect();
+        benchmark.assertRunningTime(2500);
+        benchmark.assertCounters(1000, 23, 30, 83);
+    }
+
+    @Test
+    public void testValuePredicateFiltering() {
+        String query = "match\n" +
+                "$p isa person, has $email; $email = \"douglas.schmidt@vaticle.com\";\n" +
+                "$f isa file, has $path; $path = \"root/engineering/typedb-studio/src/README.md\";\n" +
+                "$o isa operation, has $operation; $operation = \"edit file\";\n" +
+                "$a (object: $f, action: $o) isa access;\n" +
+                "$pe (subject: $p, access: $a) isa permission, has validity true;\n";
+        Benchmark benchmark = new Benchmark("value-predicate-filtering", query, 1);
+        benchmarker.runBenchmark(benchmark);
+
+        benchmark.assertAnswerCountCorrect();
+        benchmark.assertRunningTime(2500);
+        benchmark.assertCounters(1000, 200, 500, 1500);
```

Comment:
1500 compound streams to find 1 answer in 2.5 seconds! Seems like there's gotta be a better way to do this one too. 

I hope you start to see the point of why we wanted these!

Reply:
The number of answers is in no way correlated to the amount of work to be done.

This query is equivalent to "Is there a path between A and B?" which has one answer, but requires a traversal of the whole graph.

Reply:
It's true, but maybe there's a more optimal way to find that 1 answer

Reply:
Possibly - the planner picks the best plan to compute a fraction of all answers. We could also see it as computing a multiple of a a single answer.

But in this case, I think it's somewhat unlikely there's a significantly better way to do it.